# Diricore analysis

> **_IMPORTANT:_** `$BASE_DIR` has to be specified in your `.bash_profile`. Details [here](docs/0_before_you_start.md)

> **_IMPORTANT:_** Before running this analysis, python2 environment has to be set up. Details [here](docs/0_before_you_start.md). If it's set up, activate the environment with `conda activate diricore`


## 1. Create metafiles

For diricore analysis we need 2 input files - contrasts and samplenames.

1. Create samplenames file: `vim $BASE_DIR/20910/analysis/input/metadata/rpf_density_samplenames.tsv`

```
S1_T2_late      S1_T2_late      #cc00cc
S2_T4_late      S2_T4_late      #668cff
S3_T5_late      S3_T5_late      #ff6666
S4_4T1_Ctrl     S4_4T1_Ctrl     #00cc99
S5_T_Pool2      S5_T_Pool2      #262626
S6_Liver2       S6_Liver2       #b366ff
S7_Liver4       S7_Liver4       #66ffff
S8_Liver6       S8_Liver6       #862d59
S9_Liver8       S9_Liver8       #663300
S10_T_Pool1     S10_T_Pool1     #4d0000
S11_Lung2       S11_Lung2       #558000
S12_Lung4       S12_Lung4       #4d0066
S13_Lung6       S13_Lung6       #005580
S14_Lung8       S14_Lung8       #ff0000
```

2. Create contrasts file: `vim $BASE_DIR/20910/analysis/input/metadata/rpf_density_contrasts.tsv`

```
S1_T2_late      S4_4T1_Ctrl     #cc00cc
S2_T4_late      S4_4T1_Ctrl     #668cff
S3_T5_late      S4_4T1_Ctrl     #ff6666
S5_T_Pool2      S4_4T1_Ctrl     #262626
S6_Liver2       S4_4T1_Ctrl     #b366ff
S7_Liver4       S4_4T1_Ctrl     #66ffff
S8_Liver6       S4_4T1_Ctrl     #862d59
S9_Liver8       S4_4T1_Ctrl     #663300
S10_T_Pool1     S4_4T1_Ctrl     #4d0000
S11_Lung2       S4_4T1_Ctrl     #558000
S12_Lung4       S4_4T1_Ctrl     #4d0066
S13_Lung6       S4_4T1_Ctrl     #005580
S14_Lung8       S4_4T1_Ctrl     #ff0000
```

## 2. Run subsequence analysis
### 2.1 Plot creation

Subsequence analysis:  

```
bsub -q long -R "rusage[mem=50G]" $BASE_DIR/software/diricore/subsequence_analysis.sh 20910 mm10 50 all_unique
``` 
The parameters are the following:
* `20910` is the name of the dataset
* `mm10` is the reference genome. Other options: `hg19` or `yeast`
* `50` is minimal number of reads per gene. Any positive integer. Default is 100. 
* `all_unique` means that the script will analyze bam files without duplicated reads. Another option: `all` - reads with duplicates.

The script will create a data file: `ls -lh $BASE_DIR/20910/analysis/output/subsequence_data/20910.subsequence_data.all.dedup.50.hdf5`

Additionally, it will generate the plots: `$BASE_DIR/20910/analysis/output/figures/subsequence_shift_plots/all_unique`

For each contrast from the `$BASE_DIR/20910/analysis/input/metadata/rpf_density_contrasts.tsv` file, it will generate 4 plots: pos 9, pos 12, pos 15 and pos 18. 

The plot will look like that:

![](/pics/diricore_1.png)

### 2.2 Table creation

To obtain value tables for the subsequence plots, use the following bash scripts for job submission instead:

```
bsub -q long -R "rusage[mem=50G]" /omics/groups/OE0532/internal/Alex/scripts/subsequence_analysis_tables.sh 20910 mm10 50 all_unique
bsub -q long -R "rusage[mem=50G]" /omics/groups/OE0532/internal/Alex/scripts/subsequence_analysis_Mito_tables.sh 20910 mm10 50 all_unique
bsub -q long -R "rusage[mem=50G]" /omics/groups/OE0532/internal/Alex/scripts/subsequence_analysis_subset_tables.sh 20910 mm10 50 all_unique
```
With 1) General tables for all input genes; 2) Tables for mitochondrial genes; 3) Tables for subset gene input. These scripts create the subsequence plots and value tables. 
Please use the option 'plots_only' if a corresponding hdf5 file was already created in prior.


## 3. Adjust the layout

We can see that the image is a bit out of scale. We can ajust the scale with the following command: 

```
bsub -q long -R "rusage[mem=50G]" $BASE_DIR/software/diricore/subsequence_analysis.sh 20910 mm10 50 all_unique all_samples plots_only 4.0
```

where `plots_only` indicates that the script should not generate a new data file, only the plots. And `4.0` is the limits of y-axis. 

Looks much better now:

![](/pics/diricore_2.png)

## 4. Run RFP density analysis

RPF density: 

```
bsub -q long -R "rusage[mem=50G]" $BASE_DIR/software/diricore/rpf_density_analysis.sh 20910 mm10 50 all_unique
``` 

where `20910` is the name of the dataset, `mm10` is the reference genome, `50` is minimal number of reads per gene, `all_unique` means that the script will analyze bam files without duplicated reads.

The script will create a data file: `ls -lh $BASE_DIR/20910/analysis/output/rpf_5p_density/20910.txcoord_counts.all.dedup.50.hdf5`

For each amino acid it will generate a plot, all contrasts will be plotted in one image. 

The plot looks like that:

![](/pics/diricore_3.png)


## 5. Transcript distribution 

>**_NOTE_** This step can be done AFTER the RPF density analysis has ben completed. This will use the data file generated by the `rpf_density` script

Input file is required: `$BASE_DIR/20910/analysis/output/rpf_5p_density/20910.txcoord_counts.all.dedup.50.hdf5`

Generate the plot: 

```
bsub -R "rusage[mem=30G]" $BASE_DIR/software/diricore/plot_rpf_transcript_distribution.sh 20910 mm10 50 all_unique
```

The result will be written to the file: `$BASE_DIR/20910/analysis/output/figures/rpf_transcript_distribution/all_unique.20910.10.rpf_transcript_distribution_plot.20910.pdf`

As we can see, there are too many samples on one plot which makes it not so easy to understand

![](/pics/diricore_4.png)

## 6. Plot transcript distribution for a subset of samples

As we have 3 pools of samples in one dataset, we can plot each of them separately.

For each pool we need to create a list of samples which we want to be plotted together. 

Pool 1 samples: `vim $BASE_DIR/20910/analysis/input/metadata/rpf_density_samplenames_pool1.tsv`

Pool 2 samples: `vim $BASE_DIR/20910/analysis/input/metadata/rpf_density_samplenames_pool2.tsv`

Pool 3 samples: `vim $BASE_DIR/20910/analysis/input/metadata/rpf_density_samplenames_pool3.tsv`

We can create the lists manually, or just cat a part of the main `rpf_density_samplenames.tsv` file.

Pool 1: 

```
head -n 5 $BASE_DIR/20910/analysis/input/metadata/rpf_density_samplenames.tsv > 
          $BASE_DIR/20910/analysis/input/metadata/rpf_density_samplenames_pool1.tsv
```

Pool 3: 

```
tail -n 5 $BASE_DIR/20910/analysis/input/metadata/rpf_density_samplenames.tsv > 
          $BASE_DIR/20910/analysis/input/metadata/rpf_density_samplenames_pool3.tsv
```

Pool 2: 

```
head -n 10 $BASE_DIR/20910/analysis/input/metadata/rpf_density_samplenames.tsv | tail -n 5 > 
           $BASE_DIR/20910/analysis/input/metadata/rpf_density_samplenames_pool2.tsv
```

For example, Pool 1 looks like that: 

```
S1_T2_late      S1_T2_late      #cc00cc
S2_T4_late      S2_T4_late      #668cff
S3_T5_late      S3_T5_late      #ff6666
S4_4T1_Ctrl     S4_4T1_Ctrl     #00cc99
S5_T_Pool2      S5_T_Pool2      #262626
```

Then we can plot pool 1: 

```
bsub -R "rusage[mem=30G]" $BASE_DIR/software/diricore/plot_rpf_transcript_distribution.sh 20910 mm9 10 all_unique pool1
```

The script takes a subset name as an argument, which is `pool1` and looks for a `rpf_density_samplenames_pool1.tsv` file.

The result looks like that:

![](/pics/diricore_5.png)

## 6. Plot RPF density for a subset of samples

The same we can do for RPF density analysis. 

For this we need to create a file with the contrasts:

```
vim $BASE_DIR/20910/analysis/input/metadata/rpf_density_contrasts_pool1.tsv
```

And run the rpf density script: 

```
bsub -q long -R "rusage[mem=50G]" $BASE_DIR/software/diricore/rpf_density_analysis.sh 20910 mm10 10 all_unique pool1 plots_only
```

The script will create file in the same directory: `$BASE_DIR/20910/analysis/output/figures/rpf_5p_density_plots/all_unique`

The files will be named `20910.pool1.all.m10.Ala.rpf_5p_density_shift_plot.pdf`, where `pool1` is the name of our subset. 

Then only cotransts from `rpf_density_contrasts_pool1.tsv` file will be plotted together:

![](/pics/diricore_6.png)


## 7. Diricore plots with the opiton `plots_only`

Let's say, we have already a data file `.hdf5`. We just want to generate plots without generating the data file.

For all samples the commands will be:

```
bsub -q long -R "rusage[mem=50G]" $BASE_DIR/software/diricore/rpf_density_analysis.sh 20910 mm10 10 all_unique all_samples plots_only
bsub -q long -R "rusage[mem=50G]" $BASE_DIR/software/diricore/subsequence_analysis.sh 20910 mm10 50 all_unique all_samples plots_only
```

Or for example, we decided that we want to plot different contrasts. E.g. all samples vs Ctrl. So we created a new samplenames file:

```
cp $BASE_DIR/20910/analysis/input/metadata/rpf_density_samplenames.tsv $BASE_DIR/20910/analysis/input/metadata/rpf_density_samplenames_all_vs_Ctrl.tsv 
```

and we created a new contrasts file:

```
vim $BASE_DIR/20910/analysis/input/metadata/rpf_density_contrasts_all_vs_Ctrl.tsv 
```

Now we can generate diricore plots for our new contrasts:

```

bsub -q long -R "rusage[mem=50G]" $BASE_DIR/software/diricore/rpf_density_analysis.sh 20910 mm10 10 all_unique all_vs_Ctrl plots_only
bsub -q long -R "rusage[mem=50G]" $BASE_DIR/software/diricore/subsequence_analysis.sh 20910 mm10 50 all_unique all_vs_Ctrl plots_only
```

where `all_vs_Ctrl` is a contrast name (instead of `all_samples`)
